drivers/spi/spi_silabs_eusart.c

```diff
@@ -152,6 +232,326 @@ static int spi_silabs_eusart_configure(const struct device *dev, const struct sp
+	}
+
+	/* For each buffer in the instance context buffer set */
+	for (size_t i = 0U; i < buffer_count; i++) {
+		size_t buffer_remaining_count = current_buffer->len;
+
+		current_total_transfer_size += current_buffer->len;
+
+		if (buffer_remaining_count == 0U) {
+			/* Skip zero length buffers */
+			continue;
+		}
+
+		/* Create the required number of dma transfer block for this buffer */
+		while (dma_block_config_idx <= SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+			if (dma_block_config_idx >= SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+				LOG_ERR("DMA channel transfer blocks exausted. Can't prepare the "
+					"transaction.");
+				return -ENOMEM;
+			}
+
+			uint32_t index_in_buffer = current_buffer->len - buffer_remaining_count;
+
+			buffer_remaining_count = spi_silabs_build_dma_transfer_block(
+				instance_conf, &(current_blk_cfg[dma_block_config_idx]),
+				&(((uint8_t *)current_buffer->buf)[index_in_buffer]),
+				buffer_remaining_count, is_tx);
+
+			dma_block_config_idx++;
+
+			if (buffer_remaining_count == 0U) {
+				current_buffer++;
+				break;
+			}
+		}
+	}
+
+	padding_size = total_padded_transaction_size - current_total_transfer_size;
+
+	while (padding_size) {
+		if (dma_block_config_idx >= SPI_DMA_MAX_DESCRIPTOR_PER_CHANNEL) {
+			LOG_ERR("DMA channel transfer blocks exausted. Can't prepare the "
+				"transaction.");
+			return -ENOMEM;
+		}
+
+		padding_size = spi_silabs_build_dma_transfer_block(
+			instance_conf, &current_blk_cfg[dma_block_config_idx], NULL, padding_size,
+			is_tx);
+		dma_block_config_idx++;
+	}
+
+	spi_silabs_build_dma_config(dev, &dma_cfg,
+				    (const struct dma_block_config *)&current_blk_cfg[0],
+				    dma_block_config_idx, is_tx);
+
+	uint32_t dma_channel = 0U;
+	const struct device *dma_dev = NULL;
+
+	if (is_tx) {
+		dma_channel = instance_data->dma_channel_tx.channel_nb;
+		dma_dev = instance_data->dma_channel_tx.dma_dev;
+		dma_cfg.dma_slot = instance_data->dma_channel_tx.dma_slot;
+	} else {
+		dma_channel = instance_data->dma_channel_rx.channel_nb;
+		dma_dev = instance_data->dma_channel_rx.dma_dev;
+		dma_cfg.dma_slot = instance_data->dma_channel_tx.dma_slot;
+	}
+
+	ret = dma_config(dma_dev, dma_channel, &dma_cfg);
+
+	if (ret == -ENOMEM) {
+		LOG_ERR("DMA driver failed allocation.");
+		return ret;
+	}
+
+	return ret;
+}
+
+static int spi_silabs_prepare_dma_transaction(const struct device *dev,
+					      size_t total_padded_transaction_size)
+{
+	int ret;
+
+	if (total_padded_transaction_size == 0) {
+		/* Nothing to do */
+		return 0U;
+	}
+
+	ret = spi_silabs_prepare_dma_transfer(dev, total_padded_transaction_size, true);
+
+	if (ret) {
+		return ret;
+	}
+
+	ret = spi_silabs_prepare_dma_transfer(dev, total_padded_transaction_size, false);
+
+	return ret;
+}
+
+void spi_silabs_dma_rx_callback(const struct device *dev, void *user_data, uint32_t channel,
+				int status)
+{
+	if (status == DMA_STATUS_COMPLETE || status < 0U) {
```

Comment:
I suggest to handle the the exception case first:
```
if (status >= 0 && status != DMA_STATUS_COMPLETE) {
    return;
}
```
Thus, the function is simpler and you save one level of indentation.